src/tqec/compile/block.py

```diff
@@ -0,0 +1,141 @@
+    """Represents a collection of `CompiledBlock`s tiled in a 2D grid."""
+
+    def __init__(self, blocks_by_position: dict[Position2D, CompiledBlock]):
+        templates_by_position: dict[Position2D, RectangularTemplate] = {}
+        for pos, block in blocks_by_position.items():
+            templates_by_position[pos] = block.template
+        self._tiled_template = TiledTemplate(template_by_position=templates_by_position)
+        self._tiled_layers = self._merge_layers(
+            {pos: block.layers for pos, block in blocks_by_position.items()}
+        )
+
+    @property
+    def block_size(self) -> int:
+        return 2 * self._tiled_template.tile_shape.x
+
+    @property
+    def tiled_template(self) -> TiledTemplate:
+        return self._tiled_template
+
+    @property
+    def tiled_layers(self) -> list[Plaquettes]:
+        return self._tiled_layers
+
+    def _merge_layers(
+        self, layers_by_position: dict[Position2D, list[Plaquettes]]
+    ) -> list[Plaquettes]:
+        """Merge the layers of the different tiled blocks."""
+        # Check if all the blocks have the same number of layers.
+        num_layers = {len(layers) for layers in layers_by_position.values()}
+        if len(num_layers) != 1:
+            raise TQECException(
+                "All blocks in the TiledBlocks should have the same number of layers."
+            )
+        # Check if all the block layers have the same repeating structure.
+        # And merge the layers.
+        indices_map = self._tiled_template.get_indices_map_for_instantiation()
+        tiled_layers: list[Plaquettes] = []
+        for i in range(num_layers.pop()):
+            merged_plaquettes: defaultdict[int, Plaquette] = defaultdict(
+                empty_square_plaquette
+            )
+            repetitions: set[LinearFunction] = set()
+            repeated: bool = False
+            for pos, layers in layers_by_position.items():
+                layer = layers[i]
+                if isinstance(layer, RepeatedPlaquettes):
+                    repetitions.add(layer.repetitions)
+                    repeated = True
+                else:
+                    repetitions.add(LinearFunction(0, 0))
+                if len(repetitions) > 1:
+                    raise TQECException(
+                        "All the block layers on the same z-plane must have the same repeating structure, i.e.",
+                        "either all the layers are repeated and have the same scaling behavior for the repetitions"
+                        "or none of them are repeated.",
+                    )
+                imap = indices_map[pos]
+                merged_plaquettes.update(
+                    {imap[i]: plaquette for i, plaquette in layer.collection.items()}
+                )
+            plaquettes = Plaquettes(merged_plaquettes)
+            if repeated:
+                plaquettes = plaquettes.repeat(repetitions.pop())
+            tiled_layers.append(plaquettes)
```

Comment:
I prefer to use an optional here, but this might be very subjective. I am fine with keeping the set-approach if you think it is better.

```suggestion
            repetitions: LinearFunction | None = None
            for pos, layers in layers_by_position.items():
                layer = layers[i]
                if isinstance(layer, RepeatedPlaquettes):
                    if repetitions is not None and layer.repetitions != repetitions:
                        raise TQECException(
                            "All the block layers on the same z-plane must have the "
                            "same repeating structure, i.e. either all the layers are "
                            "repeated and have the same scaling behavior for the "
                            "repetitions or none of them are repeated."
                        )
                    repetitions = layer.repetitions
                imap = indices_map[pos]
                merged_plaquettes.update(
                    {imap[i]: plaquette for i, plaquette in layer.collection.items()}
                )
            plaquettes = Plaquettes(merged_plaquettes)
            if repetitions is not None:
                plaquettes = plaquettes.repeat(repetitions)
            tiled_layers.append(plaquettes)
```